stellargraph/layer/appnp.py

```diff
@@ -0,0 +1,244 @@
+        generator (FullBatchNodeGenerator): an instance of FullBatchNodeGenerator class constructed on the graph of interest
+        bias (bool): toggles an optional bias in fully connected layers
+        dropout (float): dropout rate applied to input features of each layer
+        kernel_regularizer (str): normalization applied to the kernels of fully connetcted layers
+        transport_probability: "probability" of returning to the starting node in the propogation step as desribed  in
+        the paper (alpha in the paper)
+        approx_iter: number of iterations to approximate PPNP as described in the paper (K in the paper)
+    """
+
+    def __init__(
+            self,
+            layer_sizes,
+            activations,
+            generator,
+            bias=True,
+            dropout=0.0,
+            transport_probability=0.1,
+            kernel_regularizer=None,
+            approx_iter=10
+    ):
+
+        if not isinstance(generator, FullBatchNodeGenerator):
+            raise TypeError("Generator should be a instance of FullBatchNodeGenerator")
+
+        assert len(layer_sizes) == len(activations)
+
+        self.layer_sizes = layer_sizes
+        self.transport_probability = transport_probability
+        self.activations = activations
+        self.bias = bias
+        self.dropout = dropout
+        self.kernel_regularizer = kernel_regularizer
+        self.generator = generator
+        self.support = 1
+        self.method = generator.method
+
+        # Check if the generator is producing a sparse matrix
+        self.use_sparse = generator.use_sparse
+        if self.method == "none":
+            self.graph_norm_layer = GraphPreProcessingLayer(
+                num_of_nodes=self.generator.Aadj.shape[0]
+            )
+
+        # Initialize a stack of fully connected layers
+        n_layers = len(self.layer_sizes)
+        self._layers = []
+        for ii in range(n_layers):
+            l = self.layer_sizes[ii]
+            a = self.activations[ii]
+            self._layers.append(Dropout(self.dropout))
+            self._layers.append(
+                Dense(
+                    l,
+                    activation=a,
+                    use_bias=self.bias,
+                    kernel_regularizer=self.kernel_regularizer,
+                )
+            )
+
+        feature_dim = self.layer_sizes[-1]
+
+        for ii in range(approx_iter):
+            self._layers.append(Dropout(self.dropout))
+            #use graph convolutions to perform the AH matrix multiplication (A = adjacency matrix,
+            # H = previous layer feature matrix)
+            #set the weights W = I (identity), bias=0, activation=linear, and make the layer non-trainable
+            # so the convolution becomes: activation(AHW + bias) = (AHI + 0) = AH
+            #add propogation step with lambda function in the __call__ function
+            self._layers.append(
+                GraphConvolution(
```

Comment:
Why are you using a graph convolutional layer here but not in PPNP?  It seems a bit heavyweight to use a convolutional layer instead of creating a new matrix multiplication layer, especially as you need to then append another layer to add the transport vector.  I would recommend creating a new Layer similar to the `GraphConvolutionalLayer` that does the computations you require for each APPNP iteration.

Reply:
Good idea, I'll create a new matrix multiplication layer.